patches/tModLoader/Terraria/ModLoader/BossBarLoader.cs

```diff
@@ -0,0 +1,326 @@
+		}
+
+		internal static void AddGlobalBossBar(GlobalBossBar globalBossBar) {
+			globalBossBars.Add(globalBossBar);
+			ModTypeLookup<GlobalBossBar>.Register(globalBossBar);
+		}
+
+		internal static void AddBossBarStyle(ModBossBarStyle bossBarStyle) {
+			lock (bossBarStyles) {
+				bossBarStyles.Add(bossBarStyle);
+				ModTypeLookup<ModBossBarStyle>.Register(bossBarStyle);
+			}
+		}
+
+		/// <summary>
+		/// Sets the pending style that should be switched to
+		/// </summary>
+		/// <param name="bossBarStyle">Pending boss bar style</param>
+		internal static void SwitchBossBarStyle(ModBossBarStyle bossBarStyle) => switchToStyle = bossBarStyle;
+
+		/// <summary>
+		/// Sets the saved style that should be switched to, handles possibly unloaded/invalid ones and defaults to the vanilla style
+		/// </summary>
+		internal static void GotoSavedStyle() {
+			switchToStyle = bossBarStyles.SingleOrDefault(m => m.FullName == lastSelectedStyle) ?? vanillaStyle;
+			styleLoading = false;
+		}
+
+		/// <summary>
+		/// Checks if the style was changed and applies it, saves the config if required
+		/// </summary>
+		internal static void HandleStyle() {
+			if (switchToStyle != null && switchToStyle != CurrentStyle) {
+				CurrentStyle.OnDeselected();
+				CurrentStyle = switchToStyle;
+				CurrentStyle.OnSelected();
+			}
+			switchToStyle = null;
+
+			if (!styleLoading && CurrentStyle.FullName != lastSelectedStyle) {
+				lastSelectedStyle = CurrentStyle.FullName;
+				Main.SaveSettings();
+			}
+		}
+
+		/// <summary>
+		/// Returns the texture that the given bar is using. If it does not have a custom one, it returns the vanilla texture
+		/// </summary>
+		/// <param name="bossBar">The ModBossBar</param>
+		/// <returns>Its texture, or the vanilla texture</returns>
+		public static Asset<Texture2D> GetTexture(ModBossBar bossBar) {
+			if (!bossBarTextures.TryGetValue(bossBar.index, out Asset<Texture2D> texture))
+				texture = VanillaBossBarTexture;
+			return texture;
+		}
+
+		/// <summary>
+		/// Gets the ModBossBar associated with this NPC
+		/// </summary>
+		/// <param name="npcType">The NPC type</param>
+		/// <param name="value">When this method returns, contains the ModBossBar associated with the specified NPC type</param>
+		/// <returns><see langword="true"/> if ModBossBar exists; otherwise, <see langword="false"/>.</returns>
+		public static bool NpcToBossBar(int npcType, out ModBossBar value) {
+			value = null;
+			if (npcToBossBarType.TryGetValue(npcType, out int index))
+				value = bossBars[index];
+			return value != null;
+		}
+
+		/// <summary>
+		/// Matches NPC types to their ModBossBar
+		/// </summary>
+		internal static void BuildReverseLookup() {
+			for (int i = 0; i < bossBars.Count; i++) {
+				ModBossBar bossBar = bossBars[i];
+				var validNPCs = bossBar.ValidNPCs;
+
+				//The same NPC can exist in multiple bars. In that case, only take the first found occurence
```

Comment:
this seems problematic? Is there any way for a player to select which boss bar they want for a given NPC? At the very least there should be a log message.

Reply:
Resolved in https://github.com/tModLoader/tModLoader/pull/1288/commits/e66364ddb45fdd7340b7ffe07f9fafb438288963